File: sdks/python/authgent/adapters/mcp.py
"""MCP Auth Provider adapter — plug authgent into FastMCP servers."""

from __future__ import annotations

from authgent.client import AgentAuthClient
from authgent.verify import verify_token
from authgent.models import AgentIdentity


class AgentAuthProvider:
    """Auth provider adapter for FastMCP servers.

    Usage:
        from authgent.adapters.mcp import AgentAuthProvider
        mcp = FastMCP("my-server")
        mcp.auth_provider = AgentAuthProvider(server_url="http://localhost:8000")
    """

    def __init__(self, server_url: str, audience: str | None = None):
        self._server_url = server_url.rstrip("/")
        self._audience = audience
        self._client = AgentAuthClient(server_url)

    async def verify(self, token: str) -> AgentIdentity:
        """Verify an access token and return the agent identity."""
        return await verify_token(
            token=token,
            issuer=self._server_url,
            audience=self._audience,
        )

    @property
    def metadata_url(self) -> str:
        """URL for OAuth server metadata discovery."""
        return f"{self._server_url}/.well-known/oauth-authorization-server"

    @property
    def jwks_url(self) -> str:
        """URL for JWKS endpoint."""
        return f"{self._server_url}/.well-known/jwks.json"

File: sdks/typescript/src/adapters/mcp.ts
/**
 * MCP Auth Provider adapter — plug authgent into MCP servers.
 *
 * @example
 * ```ts
 * import { AgentAuthProvider } from "authgent/adapters/mcp";
 *
 * const authProvider = new AgentAuthProvider({
 *   serverUrl: "http://localhost:8000",
 * });
 *
 * // In your MCP server handler:
 * const identity = await authProvider.verify(token);
 * console.log(identity.subject, identity.scopes);
 * ```
 */

import { verifyToken } from "../verify.js";
import type { AgentIdentity } from "../models.js";
import type { JWKSFetcher } from "../jwks.js";

export interface AgentAuthProviderOptions {
  /** URL of the authgent server. */
  serverUrl: string;
  /** Expected audience. */
  audience?: string;
  /** Custom JWKS fetcher. */
  jwksFetcher?: JWKSFetcher;
}

export class AgentAuthProvider {
  private readonly serverUrl: string;
  private readonly audience?: string;
  private readonly jwksFetcher?: JWKSFetcher;

  constructor(options: AgentAuthProviderOptions) {
    this.serverUrl = options.serverUrl.replace(/\/+$/, "");
    this.audience = options.audience;
    this.jwksFetcher = options.jwksFetcher;
  }

  /** Verify an access token and return the agent identity. */
  async verify(token: string): Promise<AgentIdentity> {
    return verifyToken({
      token,
      issuer: this.serverUrl,
      audience: this.audience,
      jwksFetcher: this.jwksFetcher,
    });
  }

  /** URL for OAuth server metadata discovery. */
  get metadataUrl(): string {
    return `${this.serverUrl}/.well-known/oauth-authorization-server`;
  }

  /** URL for JWKS endpoint. */
  get jwksUrl(): string {
    return `${this.serverUrl}/.well-known/jwks.json`;
  }
}
